mrmustard/lab_dev/states/base.py

```diff
@@ -648,6 +636,38 @@ def __init__(
         if representation is not None:
             self._representation = representation
 
+    def auto_shape(self, max_prob=0.999, max_shape=50, respect_fock_shape=True) -> tuple[int, ...]:
+        r"""
+        A good enough estimate of the Fock shape of this DM, defined as the shape of the Fock
+        array (batch excluded) if it exists, and if it doesn't exist it is computed as the shape
+        that captures at least ``settings.AUTOSHAPE_PROBABILITY`` of the probability mass of each
+        single-mode marginal (default 99.9%).
+        If the ``respect_fock_shape`` flag is set to ``True``, auto_shape will respect the
+        non-None values in ``fock_shape``.
+
+        Args:
+            max_prob: The maximum probability mass to capture in the shape (default 0.999).
+            max_shape: The maximum shape to compute (default 50).
+            respect_fock_shape: Whether to respect the non-None values in ``fock_shape``.
+        """
+        try:  # fock
+            shape = self._representation.array.shape[1:]
+        except AttributeError:  # bargmann
+            repr = self.representation
+            A, b, c = repr.A[0], repr.b[0], repr.c[0]
+            repr = repr / trace_dm(A, b, c)
```

Comment:
Out of curiosity, why not use the probability attribute of the state?

Reply:
because `.probability` triggers a long sequence of events: it is computed via the L2_norm i.e. `self >> self.dual` which calls `__rshift__`, which calls `__matmul__` which calls `gaussian_integral`. This way we have a direct fasg formula instead.

Reply:
does this mean we can generally compute probability via representation faster? should we capture this somewhere in the Bargmann class?

Reply:
I don't think it's necessary for now to run after that